tests/test_aggregates.py

```diff
@@ -0,0 +1,253 @@
+               list(aggregate_instance)
+
+    def test_valid_lambda_aggregator_function(self, setUp, project):
+        aggregate_instance = aggregator(lambda jobs: [jobs])
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+
+        assert [tuple(project)] == list(aggregate_instance)
+
+    def test_valid_sort_by(self, setUp, project):
+        helper_sort = partial(sorted, key=lambda job: job.sp.i)
+        aggregate_instance = aggregator(sort_by='i')
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+
+        assert [tuple(helper_sort(project))] == list(aggregate_instance)
+
+    def test_valid_reversed_sort(self, setUp, project):
+        helper_sort = partial(sorted, key=lambda job: job.sp.i, reverse=True)
+        aggregate_instance = aggregator(sort_by='i', reverse_order=True)
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+        assert [tuple(helper_sort(project))] == list(aggregate_instance)
+
+    def test_groups_of_valid_num(self, setUp, project):
+        valid_values = [1, 2, 3, 6, 10]
+        expected_aggregates = [10, 5, 4, 2, 1]
+
+        for i, valid_value in enumerate(valid_values):
+            aggregate_instance = aggregator.groupsof(valid_value)
+            aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+            assert len(aggregate_instance) == expected_aggregates[i]
+
+    def test_groupby_with_valid_string_key(self, setUp, project):
+        aggregate_instance = aggregator.groupby('even')
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+        for aggregate in aggregate_instance:
+            even = aggregate[0].sp.even
+            for job in aggregate:
+                assert even == job.sp.even
+        assert len(aggregate_instance) == 2
+
+    def test_groupby_with_invalid_string_key(self, setUp, project):
+        aggregate_instance = aggregator.groupby('invalid_key')
+        with pytest.raises(KeyError):
+            # We will attempt to generate aggregates but will fail in
+            # doing so due to the invalid key
+            aggregate_instance._create_AggregatesStore(project)
+
+    def test_groupby_with_default_key_for_string(self, setUp, project):
+        aggregate_instance = aggregator.groupby('half', default=-1)
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+        for aggregate in aggregate_instance:
+            half = aggregate[0].sp.get('half', -1)
+            for job in aggregate:
+                assert half == job.sp.get('half', -1)
+        assert len(aggregate_instance) == 6
+
+    def test_groupby_with_Iterable_key(self, setUp, project):
+        aggregate_instance = aggregator.groupby(['i', 'even'])
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+        # No aggregation takes place
```

Comment:
Thanks for the comment! Could you add how that means we don't need to check whether all the aggregate members are equivalent. Of course this is true for aggregates of size one, but just to be clear for future developers.